Added support for I/O from zipfiles and buffers

pfebrer opened the pull request:
#### Motivation

When dealing with very big datasets (e.g. ML) you often need to compress it into a zip or tar file to move it around. Needing to then decompress it for reading the files is a pain, because:

1. Decompressing the full dataset might take a very long time
2. File number limitations in HPC clusters. (also recently I've used a cluster that removes unused files every 30 days, which can basically destroy the parts of the dataset that you have not used).

#### Why zip and not tar?

Because tar doesn't keep an index of the contained files, so it can be very slow to access files within it when the tarfile gets big.

Therefore I have determined that putting effort into zip files is much more worth it.

#### Approach taken in this PR

Turns out that the `zipfile` library already had a `zipfile.Path` object that mimics `pathlib.Path`. Therefore, `zipfile.Path` can be used within the sile framework. It just needed some extra functionality to fully work, so I created an extension `ZipPath`, which:

- Implements missing `pathlib.Path` functions that the sile framework was using, e.g. `with_suffix`.
- Wraps the `open` method of the returned file handles so that on close we can close also the root zip file. This is specially important when writing.

**What works at the moment:**
- Reading and writing text based files.
- Reading cdf files. This is done through the `memory` argument of `netCDF4.Dataset`, which accepts the in memory bytes.
- Reading binary files. This is done by writing on a temporary file. I need some clever way of knowing when the temporary file can be removed though.
- Searching for other files in the directory. E.g. when reading from an fdf inside a zipfile, it can find other files (other fdfs, basis files, matrices) that are also inside the zipfile.

**What doesn't work for now:**
- Writing CDF files. This can also be done through the `memory` argument.
- Writing binary files. This can also be done through temporary files.

#### Comments

I will implement the missing parts and also add tests when I get feedback on whether this approach is fine or it is preferrable to create separate sile classes as with `BufferSile`.

The handling of CDF and binary files I think it should be generalized to handle generic python filehandles/objects (e.g. `bytes`/`BytesIO`) and not just contents of a zip file.

#### How to test

Here is a zip file that you can use for testing, which contains a SIESTA run of a water molecule:

[run.zip](https://github.com/user-attachments/files/19929287/run.zip)

You can read the Hamiltonian like:

```python
import sisl

sisl.get_sile("/path/to/run.zip/run/RUN.fdf").read_hamiltonian()
```

The sile framework will automatically detect that the path contains a zip file and do the necessary adjustments.

**Hope you like it!**

zerothi commented:
Gould you rebase on main, to remove basis cli?

Looks very interesting!

On Sun, Apr 27, 2025, 13:42 codecov[bot] ***@***.***> wrote:

> *codecov[bot]* left a comment (zerothi/sisl#914)
> <https://github.com/zerothi/sisl/pull/914#issuecomment-2833414647>
> Codecov
> <https://app.codecov.io/gh/zerothi/sisl/pull/914?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior>
> Report
>
> Attention: Patch coverage is 56.41026% with 34 lines in your changes
> missing coverage. Please review.
>
> Project coverage is 86.88%. Comparing base (e902a1e)
> <https://app.codecov.io/gh/zerothi/sisl/commit/e902a1e0964c264757e3718f10ecfc16ea46435a?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior>
> to head (b5e3387)
> <https://app.codecov.io/gh/zerothi/sisl/commit/b5e3387fef4ab44bb11b9a27afda9d1d97ddb2dc?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior>
> .
>
> Files with missing lines
> <https://app.codecov.io/gh/zerothi/sisl/pull/914?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior> Patch
> % Lines
> src/sisl/io/_zipfile.py
> <https://app.codecov.io/gh/zerothi/sisl/pull/914?src=pr&el=tree&filepath=src%2Fsisl%2Fio%2F_zipfile.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior#diff-c3JjL3Npc2wvaW8vX3ppcGZpbGUucHk=>
> 41.02% 23 Missing ⚠️
> <https://app.codecov.io/gh/zerothi/sisl/pull/914?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior>
> src/sisl/io/sile.py
> <https://app.codecov.io/gh/zerothi/sisl/pull/914?src=pr&el=tree&filepath=src%2Fsisl%2Fio%2Fsile.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior#diff-c3JjL3Npc2wvaW8vc2lsZS5weQ==>
> 71.79% 11 Missing ⚠️
> <https://app.codecov.io/gh/zerothi/sisl/pull/914?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior> Additional
> details and impacted files
>
> @@            Coverage Diff             @@##             main     #914      +/-   ##
> ==========================================- Coverage   86.93%   86.88%   -0.05%
> ==========================================
>   Files         408      409       +1
>   Lines       53799    53869      +70
> ==========================================+ Hits        46769    46806      +37     - Misses       7030     7063      +33
>
> ☔ View full report in Codecov by Sentry
> <https://app.codecov.io/gh/zerothi/sisl/pull/914?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior>
> .
> 📢 Have feedback on the report? Share it here
> <https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=Nick+Papior>
> .
> 🚀 New features to boost your workflow:
>
>    - ❄️ Test Analytics <https://docs.codecov.com/docs/test-analytics>:
>    Detect flaky tests, report on failures, and find test suite problems.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/zerothi/sisl/pull/914#issuecomment-2833414647>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABERMIPHT4RCCH2N2TY6FKD23S7CXAVCNFSM6AAAAAB36P2EN2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMZTGQYTINRUG4>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

pfebrer commented:
The basis commits are already removed :+1:

zerothi commented on src/sisl/io/_zipfile.py:
```diff
@@ -0,0 +1,94 @@
+
+        Given a path object, scans
+        """
+        for i, part in enumerate(path.parts[:-1]):
```
It seems the CI is correct here, the `else` is wrongly formatted, will never occur, no break is present.

pfebrer replied:
Hmm yeah ok I guess it is not 100% correct structuring. The `else` will happen though, because it happens if the end of the loop is reached.

There is no break but I exit the loop with a `return`. It is true that `else` is not needed then, we can just have the raise after the loop.

In other words, the code works but it is not the best way to do it :sweat_smile:

zerothi commented on src/sisl/io/sile.py:
```diff
@@ -497,6 +500,33 @@ class BaseSile:
+            except FileNotFoundError:
+                pass
+
+        return filename
```
it should convert to `Path` here, to follow the signature.

pfebrer replied:
It is already converted to `Path` at the beggining of the `elif` and everything should go through that `elif` except the case when filename is a `ZipPath`

zerothi replied:
true, missed that!

zerothi commented on src/sisl/io/sile.py:
```diff
@@ -1574,13 +1614,45 @@ class SileBin(BaseSile):
+        # We therefore store the contents in a temporary file and set this
+        # as the file to be read from.
+        if (
+            name.startswith("read_")
```
what about write? Should that even be allowed?

pfebrer replied:
Yes, I think so. It is a bit more scary because if we don't manage the write correctly it can end up in a corrupted zip file, that's why I didn't want to implement it fast, I need to put a bit of thought into it.

zerothi replied:
Otherwise, doing an `NotImplementedError` is sufficiently at the moment, it doesn't seem like it is something that's very useful ATM.

pfebrer replied:
Yes but on the other hand if I do it now it will be faster than if I come back at it 2 years later :sweat_smile:

zerothi reviewed (commented):
I'll likely have some more comments, but it looks very clean, simple and effective. Major Q is whether `write` works, otherwise an error should be raised.

pfebrer commented:
Yes, I think there will be no problem in making write work :+1:

pfebrer commented:
By the way, what I've done for CDF and binary files can work really with any buffer, its not like it just works with zipfiles. So I think it would be worth it now that we are at it to make it support arbitrary buffers, which is something that has come up in the past.

zerothi commented:
> By the way, what I've done for CDF and binary files can work really with any buffer, its not like it just works with zipfiles. So I think it would be worth it now that we are at it to make it support arbitrary buffers, which is something that has come up in the past.
> 
> Any ideas of how things could be structured to support the buffers? Just extend the `BufferSile` to handle more buffer types, create a new class...?

If the `BufferSile` could be extended as a baseclass for binary + text files, just like the `sile`, then I think that would be the most optimal way of doing things.

So:
```mermaid
classDiagram
    BaseSile <|-- Sile
    BaseSile <|-- SileCDF 
    BaseBufferSile
    BaseBufferSile <|-- BufferSile
    BaseBufferSile <|-- BufferCDF 
```

pfebrer commented:
Ok, will try to do that :+1:

By the way, where is the documentation for generating this diagrams? hahah

zerothi commented:
> Ok, will try to do that 👍
Nice!

> 
> By the way, where is the documentation for generating this diagrams? hahah

They are [mermaid charts](https://www.mermaidchart.com/) ;)

zerothi commented:
I think this is great!

Only one comment.

I am not sure I like the `get_sile("hello/test.zip/...")` should automatically create the `test.zip`. For reading it is fine, because you can detect whether it is a zip file, but for creation, I am a bit worried that it might be ambiguous, should it be a directory, or a zip file?
I think for creating files we shouldn't allow that nomenclature... Could one extend the `ZipPath` to have a *base* Zip-path so you could do something like:

```python
from sisl.io import ZipPath

zipfile = ZipPath("/hello/test.zip") / "dir_in_zip_file"
si.get_sile(zipfile, mode="w")
```
With the above, there should be no ambiguity.
I

pfebrer commented:
Hmm ok I understand but for writing the zip file will often also exist already, you just want to add files to It. So what we could do maybe is that `sisl` will never create a new zip file.

If you want to open the zip file externally I think the best thing to do is to use the `zipfile` classes. I think `ZipPath` should remain internal to sisl. This is what I show in the last example of the changes rst.

zerothi commented:
> Hmm ok I understand but, for writing, the zip file will often also exist already, you just want to add files to It. So what we could do maybe is that `sisl` will never create a new zip file.
If it exists, there will be no ambiguity, unless you specify `mode='w'`, in which case you want to overwrite it...

> If you want to open the zip file externally I think the best thing to do is to use the `zipfile` classes. I think `ZipPath` should remain internal to sisl. This is what I show in the last example of the changes rst.
Ok!

pfebrer commented:
Yes, when the zipfile is opened inside sisl it is always with mode "r" or "a", never "w".

zerothi commented:
> Yes, when the zipfile is opened inside sisl it is always with mode "r" or "a", never "w".

what happens when the user requests `mode='w'`?

pfebrer commented on src/sisl/io/_zipfile.py:
```diff
@@ -0,0 +1,131 @@
+        for i, part in enumerate(path.parts[:-1]):
+            if part.endswith(".zip"):
+                zip_path = Path(*path.parts[: i + 1])
+                if zip_path.is_file():
+                    zipfile_mode = "r" if mode.startswith("r") else "a"
+                    root_zip = zipfile.ZipFile(zip_path, zipfile_mode)
+                    return cls(
+                        root_zip,
+                        str(path.relative_to(zip_path)),
+                        close_zipfile=close_zipfile,
+                    )
```
@zerothi we are already checking that the zip file exists before opening it, I didn't remember :sweat_smile:

zerothi commented:
> The `mode="w"` that the user requests is for the file, not for the container zip file.
> 
> It is the same as if we wanted to write to a folder, we want to add things to the folder, never overwrite the full folder to write the file.

Hmm... Yeah, I see... :)

pfebrer commented:
Ok, so this is now super ready, I'm very happy with the final result, everything works:

**Read/write from/to buffers and zipfiles for ALL siles.** :partying_face: 

Plus all write/read operations for buffers and zipfiles are tested (both text, binary and CDF). So things are quite robust and I'm confident it should work fine.

I just have to figure out why the tests with temporary files fail in the CI (only with ubuntu). Could you try the tests in your computer and see if everything is fine?

I also refactored the I/O documentation to include information on how to do I/O with zipfiles and buffers, and added a link to it in the introduction section, because it seemed to me that it was not visible enough and it is an important part of the package :sweat_smile:

pfebrer commented:
So it is quite a mess to support zipfiles both in python 3.9 and python>3.9 because there was one key change in between: https://github.com/python/cpython/issues/84744#issuecomment-1093871456.

This means that we have to have many extra checks and code for a python version that we are going to drop in October :( The easiest fix also involves requiring an extra dependency `zipp`.

So I suggest it is not worth it to support zipfiles in python 3.9, we can just raise a meaningful error to update to python >= 3.10 if the user wants to use zipfiles. Is this ok?

zerothi commented:
> So it is quite a mess to support zipfiles both in python 3.9 and python>3.9 because there was one key change in between: [python/cpython#84744 (comment)](https://github.com/python/cpython/issues/84744#issuecomment-1093871456).
> 
> This means that we have to have many extra checks and code for a python version that we are going to drop in October :( The easiest fix also involves requiring an extra dependency `zipp`.
> 
> So I suggest it is not worth it to support zipfiles in python 3.9, we can just raise a meaningful error to update to python >= 3.10 if the user wants to use zipfiles. Is this ok?

Yes, that is fine. :)
Thanks, sorry, this week is busy up to the workshop, so I'll likely first have time to get things in after the workshop...
